crates/sui-core/src/authority_server.rs

```diff
@@ -378,11 +362,6 @@ impl ValidatorService {
             // For owned objects this will return without waiting for certificate to be sequenced
             // First do quick dirty non-async check
             if !epoch_store.is_tx_cert_consensus_message_processed(&certificate)? {
-                if !consensus_adapter.check_limits() {
```

Comment:
Rather than move the check, any reason not to have it in both places? If we are overloaded we probably want to load shed everything?

Reply:
Right now we have 3 checks for load shedding:
1. `TooManyTransactionsPendingExecution`
2. `TooManyTransactionsPendingOnObject`
3. Consensus overloaded.

My understanding is that having all three checks during the transaction signing phase should be enough for system protection during traffic surge. From user perspective it's probably cleaner as well (once a certificate is formed, the system guarantees that it's going to be executed).

There could be some attacks, in which an attacker is gathering a lot of certificates during the epoch and sending them for execution all at once: these attacks might temporarily increase latencies, as txn manager is going to have a longer execution queue (note that NW has its own semaphore protection). If we want to be able to avoid that situation, we should duplicate the very same checks both for txn signing and handling certificate.

Reply:
My thinking is, after tx is signed you could later try to execute it any time during the epoch, is that right? Even if it's hours later? If these checks are cheap (I think they are?) then we may as well apply the checks to both operations, because system load can fluctuate. If we only apply the check to one of them, then it probably makes more sense for it to be execution and not cert signing, because execution is much more expensive? And we'd rather prevent the expensive operation rather than the cheap operation, when we are overloaded. Also, applying to both stops users from getting signatures from validators that are not overloaded and then using it to force traffic onto a validator that *is* overloaded.

Curious what @andll et al think though

Reply:
I think we should apply all three checks to both signing and submission phases, just so we can worry less about someone accumulating a large number of signed transactions and send them in all at once (still problematic because gas cannot be adjusted on signed transactions).

TxManager limits are already applied to both code paths, so adding consensus adapter check to signing is a good idea.

Reply:
> you could later try to execute it any time during the epoch, is that right

Yes, that's right. My mental model was indeed about the typical flows with cert execution coming right after txn signing.
One reason for using txn signing for load shedding is that it'd better fit txn prioritization we want to introduce later to the system (based on gas price, hot object access logic). Once we prioritize some txns over the others based on gas price, that prioritization must happen at signing stage because a client cannot update gas price once the objects are locked. Hence, the idea would be that given a certificate a client has stronger guarantees about this cert execution.
However, the signing-phase load shedding doesn't prevent quasi-byzantine attacks like the one you described above (gathering multiple certificates and submitting them later).

Reply:
I think ultimately, there may need to be some restrictions on who can directly talk to validators for signing and submissions, e.g. only from fullnodes associated with any validator (validator fullnodes). Until then, it seems hard to have totally bullet proof load shedding.